### ISSUE TITLE
Expanded terminal (dual-monitor) with setting "Always on Top"

### ISSUE THREAD

**Issue by seniorihor** [noctuid/tdrop#17]
I'm curious if there's an option to make drop-down terminal be opened with option "Always on Top" at Gnome graphic environment?

**Comment by noctuid** [comment]
You can set this window property using wmctrl. After installing wmctrl, something like this should work:

```
tdrop -c 'wmctrl -i -r $1 -b add,above' xterm
```
Let me know if you encounter any problems.

**Comment by seniorihor** [comment]
I execute this command and get --help info of `tdrop` as result and nothing else happens.
The same outcome with other terminals.
Any other ideas?

P.S.: `wmctrl` and `xterm` are installed.

**Comment by noctuid** [comment]
Tdrop also needs to be updated. I just added this flag, so it doesn't exist if you have an older version installed. Let me know if that fixes the problem.

**Comment by seniorihor** [comment]
Thank you for your fast reaction!
This feature works perfectly for `xterm`, but I got troubles using it with my current terminal - [terminix](https://github.com/gnunn1/terminix).
First, it seems like option "Always on Top" isn't available for maximized windows.
If I setup `terminix` to be launched at non-maximum size it works good on first launch.
But following terminal launches makes it maximized again and it's not on top accordingly.

So, I dig a bit about solutions on side of `terminix` and found options which would fit to my needs.
But for some reason I cannot pass the command to `tdrop` this way:
`tdrop terminix --focus-window`

Is there some specific syntax about using options for terminal?

**Comment by noctuid** [comment]
> after first launch xterm starts without the option

I added a `--map-hook` that works like `-c` but will always run the command. You could possibly use this to also prevent terminix from opening maximized, but I think using `--focus-window` is probably the better idea.

> Is there some specific syntax about using options for terminal?

Yes, tdrop has `-f`/`--program-flags`. You don't need to update tdrop if you want to use this. For example: `tdrop -f '--focus-window' terminix`.

**Comment by seniorihor** [comment]
Thank you so much! :smile:
This is exactly what I need.

czw., 5 maj 2016, 18:57 użytkownik Fox Kiester <notifications@github.com>
napisał:

> after first launch xterm starts without the option
>
> I added a --map-hook that works like -c but will always run the command.
> You could possibly use this to also prevent terminix from opening
> maximized, but I think using --focus-window is probably the better idea.
>
> Is there some specific syntax about using options for terminal?
>
> Yes, tdrop has -f/--program-flags. You don't need to update tdrop if you
> want to use this. For example: tdrop -f '--focus-window' terminix.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/noctuid/tdrop/issues/17#issuecomment-217210100>
>
-- 
Sincerely yours,
Ihor Breza

**Comment by seniorihor** [comment]
I noticed `--program-flags` works only on first launch:
`tdrop --map-hook 'wmctrl -i -r $1 -b add,above' --program-flags '--focus-window --full-screen' terminix`
Which means that first time I got full-screen focused terminal and after that window is unfocused and maximized.
So, is there an alternative to `--program-flags`, which would always run the command when the dropdown is opened or other way to achieve that result?

**Comment by noctuid** [comment]
Yes, the program with its flags is only called when first creating it. In this case I guess the window needs to be focused every time it is mapped.

You may be able to use xdotool to fix the problem:
```
tdrop --map-hook 'xdotool windowactivate $1' terminix
```

**Comment by seniorihor** [comment]
OK, but it would solve just half of mine problems.
Because I'm also play around with setting specific size of `terminix`, using [`--geometry` switch](https://github.com/gnunn1/terminix/issues/300).
So I expect to run it's options on every dropdown.
What solution would you suggest for such case?

**Comment by noctuid** [comment]
> So I expect to run it's options on every dropdown.

The program is only started once, so the program flags are only used once. Is there some problem with using tdrop's `-w`, `-h`, `-x`, and `-y` to set the geometry?

Also, for dual monitors, the dropdown should span them if you use a percentage as the width (e.g. `--width "100%"`) as long as you don't use the `-m`/`--monitor-aware` flag.

**Comment by seniorihor** [comment]
> Is there some problem with using tdrop's -w, -h, -x, and -y to set the geometry?

I tried to use `tdrop` to get terminal which spans across monitors, but it looks for me not working properly with `terminix`.
Following command open window, maximized for one monitor:
```
tdrop --width "200%" --height "100%" terminix
``` 
Though, other applications (like `gedit`) expanding correctly.
So I though I can find out how to make it on side of `terminix` instead.
It would be great if I could use `tdrop` for this behavior.

**Comment by noctuid** [comment]
Since the terminix's `--geometry` option won't let you expand further than one monitor, I assume this is a problem with terminix and there is nothing I can do about it. I'll watch the issue you linked to. If after the terminix's geometry is fixed, tdrop's geometry setting flags still don't work for this, let me know.

**Comment by seniorihor** [comment]
UPDATE:
I changed profile settings for terminal size in Preferences and now the command launch `terminix` expanded for 2 monitors, but... only on first launch. :smile:

**Comment by noctuid** [comment]
Does changing the profile settings do this even if you don't use `--width` or `--height` with tdrop? You could maybe try this:

```
tdrop --width "100%" --height "100%" --map-hook 'xdotool windowactivate $1 windowsize $1 $width $height' terminix
```

If xdotool can't resize the window across the monitors, it's probably a problem with terminix.

**Comment by seniorihor** [comment]
I just decreased size in Preferences, because it was set to maximum: 511 columns, 511 rows.
If I run `terminix` without any options, it appears at size set in profile preferences.

The command you suggested works as expected on first launch:
```
tdrop --width "200%" --height "100%" --map-hook 'xdotool windowactivate $1 windowsize $1 $width $height' terminix
```
And after following dropdowns I get the error:
```
XGetWindowProperty[_NET_WM_DESKTOP] failed (code=1)
```
and maximized (not expanded) terminal window.

**Comment by noctuid** [comment]
I probably should have mentioned this earlier, but Gnome is a DE that changes the location of windows when remapping them, so you should use the `-a` flag as well.

If that doesn't do anything, could you tell me what Gnome you are using (is it Gnome shell?)  and what the output of the following is:

```
xprop -notype -id "$(xprop -root -notype | \
	awk '$1=="_NET_SUPPORTING_WM_CHECK:"{print $5}')" \
	-f _NET_WM_NAME 8u | awk -F "\"" '/WM_NAME/ {print $2}' | head -n 1 
```

Also, if `-a` doesn't work, this should in the meantime:
```
my_tdrop --x-offset <your-x> --y-offset <your-y> --width "100%" --height "40%" \
	--map-hook 'xdotool windowactivate $1 windowmove $1 $xoff $yoff windowsize $1 $width $height' terminix
```

I was able to replicate your problem on another DE, and this fixed it for me. The error you got can be ignored.

**Comment by noctuid** [comment]
Apologies if you get two notifications for this. I accidentally deleted the comment I just made.

Gnome moves windows after remapping them, so you'll need to use the `-a` flag as well in your tdrop command. If that doesn't work, could you tell me what DE you are using (is it Gnome shell?) and give me the output of the following command:
```
xprop -notype -id "$(xprop -root -notype | \
	awk '$1=="_NET_SUPPORTING_WM_CHECK:"{print $5}')" \
	-f _NET_WM_NAME 8u | awk -F "\"" '/WM_NAME/ {print $2}' | head -n 1
```

If `-a` doesn't work, you can use this as well:
```
tdrop --x-offset <xoff> --y-offset <yoff> --width "100%" --height "40%" \
	--map-hook 'xdotool windowactivate $1 windowmove $1 $xoff $yoff windowsize $1 $width $height' terminix
```

I was able to replicate your problem on a DE, and this fixed it for me. The error can be ignored.

**Comment by seniorihor** [comment]
Alright, with flag `-a` `tdrop` is pretty much useful for me and it works almost perfect on my Gnome Shell 3.20.1 (Ubuntu 16.04).
Anyway, believe there's a bug, just don't know what component cause it.
I'll try to describe it, but it'd be more clearly if you reproduce it on your own.

Command I use can't be launched by standard Gnome utility for Keyboard Shortcuts after system startup.
If I run it from command line I get error, but `terminix` appears and keybinding after that works correctly.
```
$ tdrop -a --width '200%' --height '103%' --map-hook 'wmctrl -i -r $1 -b add,above' --program-flags '--focus-window' terminix
X Error of failed request:  BadWindow (invalid Window parameter)
  Major opcode of failed request:  8 (X_MapWindow)
  Resource id in failed request:  0x0
  Serial number of failed request:  17
  Current serial number in output stream:  20
```
(here I use 103% for height, because 100% doesn't expand terminal fully)
I get identical error if I run the same command without `--width` and `--height` flags.
Though, this small fix makes keybinding work on startup, but terminal size is expanded just horizontally and not vertically.
It looks like DE remembers size and position of terminal, because after first launch from console command without flags `--width` and `--height` produces expected size.

Basically, I got it working as I wanted and this error doesn't make much troubles for me.
But it'd be nice to figure out what is reason of the fail and fix it if possible.
@noctuid Thank you for your help!

**Comment by noctuid** [comment]
I got rid of the error, but I don't think it's related to any of these problems.

> can't be launched by standard Gnome utility for Keyboard Shortcuts

I couldn't get keyboard shortcuts working at all in Gnome shell unfortunately, so I don't know why this is happening. I don't know if there is some limit to the length of commands. It seems unlikely, but you might want to try shortening the command. It shouldn't be necessary to use `--program-flags` anymore. Also, you can use '103%' for both the width and the height. Try this:

```
tdrop -a -w '103%' -h '103%' terminix
```
You can keep the wmctrl command in as well. Is it just the width and height flags that are causing a problem for the Gnome keyboard shortcut? Is there still a problem if you use an absolute number like `-w 1000`?

> DE remembers size and position of terminal

Gnome shell shouldn't alter the size of the windows, but it does change the position (which is why the `-a` is necessary).

**Comment by seniorihor** [comment]
After upgrade of `tdrop` there's no more error message in command line.
But problem with Gnome Keyboard Shortcut still remains like before, even if decrease length of command using shortened options. either using an absolute numbers
So this command works from command line, but not using Gnome keybinding:
`tdrop -a -w 2000 -h 1030 -s 'wmctrl -i -r $1 -b add,above' -f '--focus-window' terminix`

**Comment by noctuid** [comment]
I haven't been able to get Gnome's keyboard shortcuts working at all, so I'm having a hard time replicating this. Can you identify what part of the command is preventing things from working? Did you say it works fine if you remove the width and height flags?  There may not be anything I can do about this. A possible workaround would be to use another program to bind the command such as sxhkd or xbindkeys.

**Comment by seniorihor** [comment]
> Did you say it works fine if you remove the width and height flags?

Yes, it work without those flags.

OK, that's not a big problem for me anyway.
Thank you for your assistance with the configuration! :+1:

**Comment by noctuid** [comment]
I meant to close this a while ago. Let me know if there is anything else I can do/add.
